File: client-ios-swift/Messaging/DataModules/API/VoximplantService.swift
/*
*  Copyright (c) 2011-2019, Zingaya, Inc. All rights reserved.
*/

import Foundation
import VoxImplant

#error ("Enter Voximplant account credentials")
let appName = "messaging"
let accountName = "mobiledev"
let voxDomain = ".voximplant.com"

protocol VoximplantServiceDelegate: AnyObject {
    func didReceive(conversationEvent: VIConversationEvent)
    func didReceive(messageEvent: VIMessageEvent)
    func didReceive(serviceEvent: VIConversationServiceEvent)
    func didReceive(userEvent: VIUserEvent)
}

class VoximplantService: NSObject, VIMessengerDelegate, MessagingDataSource {
    private weak var delegate: VoximplantServiceDelegate?
    
    private let messenger: VIMessenger
    
    var me: VIUser?
    private var meNeedsUpdating: Bool = false
    
    init(with messenger: VIMessenger) {
        self.messenger = messenger
        super.init()
        self.messenger.addDelegate(self)
    }
    
    func set(delegate: VoximplantServiceDelegate) {
        self.delegate = delegate
    }
    
    // MARK: - Users -
    func removeMe() {
        me = nil
        meNeedsUpdating = false
    }
    
    // MARK: - Request
    func requestMe(completion: @escaping VIUserCompletion) {
        if let me = self.me,
            !meNeedsUpdating { completion(.success(me)) }
            
        else if let myUsername = messenger.me {
            requestUser(with: myUsername) { result in
                if case .failure (let error) = result { completion(.failure(error)) }
                if case .success (let user) = result {
                    self.me = user
                    self.meNeedsUpdating = false
                    completion(.success(user))
                }
            }
        }
            
        else { completion(.failure(VoxDemoError.errorNotLoggedIn())) }
    }
    
    func requestUser(with imID: NSNumber, completion: @escaping VIUserCompletion) {
        messenger.getUserByIMId(imID, completion: VIMessengerCompletion<VIUserEvent> (success:
            { userEvent in
                completion(.success(userEvent.user)) })
            { errorEvent in
                completion(.failure(NSError.buildError(from: errorEvent)))
            })
    }
    
    func requestUser(with username: String, completion: @escaping VIUserCompletion) {
        messenger.getUserByName(username, completion: VIMessengerCompletion<VIUserEvent> (success:
            { userEvent in
                completion(.success(userEvent.user)) })
            { errorEvent in
                completion(.failure(NSError.buildError(from: errorEvent)))
            })
    }
    
    func requestUsers(with usernameArray: [String], completion: @escaping VIUserArrayCompletion) { // TODO: - check for count of usernames
        messenger.getUsersByName(usernameArray, completion: VIMessengerCompletion<NSArray> (success:
            { event in
                let userEvents = event as! [VIUserEvent]
                completion(.success(userEvents.map { $0.user }))
            })
            { errorEvent in
                completion(.failure(NSError.buildError(from: errorEvent)))
            })
    }
    
    func requestUsers(with imIDArray: [NSNumber], completion: @escaping VIUserArrayCompletion) { // TODO: - check for count of usernames
        messenger.getUsersByIMId(imIDArray, completion: VIMessengerCompletion<NSArray> (success:
            { event in
                let userEvents = event as! [VIUserEvent]
                completion(.success(userEvents.map { $0.user }))
            })
            { errorEvent in
                completion(.failure(NSError.buildError(from: errorEvent)))
            })
    }
    
    // MARK: - Edit
    func editUser(with customData: [String: NSObject], completion: @escaping VIUserCompletion) {
        messenger.editUser(withCustomData: customData, privateCustomData: nil, completion: VIMessengerCompletion<VIUserEvent> (success:
            { userEvent in
                self.me = userEvent.user
                completion(.success(userEvent.user))
            })
            { errorEvent in
                completion(.failure(NSError.buildError(from: errorEvent)))
            })
    }
    
    // MARK: - Conversations -
    // MARK: - Create
    func createConversation(with config: VIConversationConfig, completion: @escaping VIConversationCompletion) { // TODO: - Discuss
        messenger.createConversation(config, completion: VIMessengerCompletion<VIConversationEvent> (success:
            { conversationEvent in
                self.meNeedsUpdating = true
                completion(.success(conversationEvent.conversation)) })
            { errorEvent in
                completion(.failure(NSError.buildError(from: errorEvent)))
            })
    }
    
    // MARK: - Request
    func requestSingleConversation(with uuid: String, completion: @escaping VIConversationCompletion) {
        messenger.getConversation(uuid, completion: VIMessengerCompletion<VIConversationEvent> (success:
            { conversation in
                completion(.success(conversation.conversation))
            })
            { errorEvent in
                completion(.failure(NSError.buildError(from: errorEvent)))
            })
    }
    
    func requestMultipleConversations(with uuids: [String], completion: @escaping VIConversationArrayCompletion) {
        messenger.getConversations(uuids, completion: VIMessengerCompletion<NSArray> (success:
            { conversationEvents in
                let conversationEvents = conversationEvents as! [VIConversationEvent]
                let conversations = conversationEvents.map { conversationEvent in conversationEvent.conversation }
                completion(.success(conversations))
            })
            { errorEvent in
                completion(.failure(NSError.buildError(from: errorEvent)))
            })
    }
    
    func recreateConversation(with UUID: String, and config: VIConversationConfig, _ lastSequence: Int?) -> VIConversation? {
        return messenger.recreateConversation(config, uuid: UUID, lastSequence: Int64(lastSequence ?? 0), lastUpdateTime: 0, createdTime: 0)
    }
    
    // MARK: - Edit
    func edit(participants: [VIConversationParticipant], in conversation: VIConversation, completion: @escaping VIConversationEventCompletion) {
        conversation.editParticipants(participants, completion: VIMessengerCompletion<VIConversationEvent> (success:
            { conversationEvent in
                completion(.success(conversationEvent))
            })
            { errorEvent in
                completion(.failure(NSError.buildError(from: errorEvent)))
            })
    }
    
    func add(participants: [VIConversationParticipant], to conversation: VIConversation, completion: @escaping VIConversationEventCompletion) {
        conversation.addParticipants(participants, completion: VIMessengerCompletion<VIConversationEvent> (success:
            { conversationEvent in
                completion(.success(conversationEvent))
            })
            { errorEvent in
                completion(.failure(NSError.buildError(from: errorEvent)))
            })
    }
    
    func remove(participants: [VIConversationParticipant], from conversation: VIConversation, completion: @escaping VIConversationEventCompletion) {
        conversation.removeParticipants(participants, completion: VIMessengerCompletion<VIConversationEvent> (success:
            { conversationEvent in
                completion(.success(conversationEvent))
            })
            { errorEvent in
                completion(.failure(NSError.buildError(from: errorEvent)))
            })
    }
    
    func update(conversation: VIConversation, completion: @escaping VIConversationEventCompletion) {
        conversation.update(completion: VIMessengerCompletion<VIConversationEvent> (success:
            { conversationEvent in
                completion(.success(conversationEvent))
            })
            { errorEvent in
                completion(.failure(NSError.buildError(from: errorEvent)))
            })
    }
    
    func leaveConversation(with UUID: String, completion: @escaping VIConversationEventCompletion) {
        messenger.leaveConversation(UUID, completion: VIMessengerCompletion<VIConversationEvent> (success:
            { conversationEvent in
                completion(.success(conversationEvent))
            })
            { errorEvent in
                completion(.failure(NSError.buildError(from: errorEvent)))
            })
    }
    
    // MARK: - Messages -
    func sendMessage(with text: String, in conversation: VIConversation, completion: @escaping VIMessageEventCompletion) {
        conversation.sendMessage(text, payload: nil, completion: VIMessengerCompletion<VIMessageEvent> (success:
            { messageEvent in
                completion(.success(messageEvent)) })
            { errorEvent in
                completion(.failure(NSError.buildError(from: errorEvent)))
            })
    }
    
    func edit(message: VIMessage, with text: String, completion: @escaping VIMessageEventCompletion) {
        message.update(text, payload: nil, completion: VIMessengerCompletion<VIMessageEvent> (success:
            { messageEvent in
                completion(.success(messageEvent))
            })
            { errorEvent in
                completion(.failure(NSError.buildError(from: errorEvent)))
            })
    }
    
    func remove(message: VIMessage, completion: @escaping VIMessageEventCompletion) {
        message.remove(completion: VIMessengerCompletion<VIMessageEvent> (success:
            { messageEvent in
                completion(.success(messageEvent))
            })
            { errorEvent in
                completion(.failure(NSError.buildError(from: errorEvent)))
            })
    }
    
    func recreateMessage(with UUID: String, and conversationUUID: String) -> VIMessage? {
        return messenger.recreateMessage(UUID, conversation: conversationUUID, text: nil, payload: nil, sequence: 0)
    }
    
    // MARK: - Events -
    func markAsRead(sequence: Int64, in conversation: VIConversation) {
        // completion is nil and event will be sent to delegate method
        conversation.markAsRead(sequence, completion: nil)
    }
    
    func sendTyping(in conversation: VIConversation) {
        // completion is nil and event will be sent to delegate method
        conversation.typing(completion: nil)
    }
    
    func requestMessengerEvents(for conversation: VIConversation, completion: @escaping VIEventArrayCompletion) {
        conversation.retransmitEvents(to: conversation.lastSequence, count: UInt(100),completion: VIMessengerCompletion<VIRetransmitEvent> (success:
            { retransmitEvents in
                completion(.success(retransmitEvents.events))
            })
            { errorEvent in
                completion(.failure(NSError.buildError(from: errorEvent)))
            })
    }
    
    // MARK: - VIMessengerDelegate -
    func messenger(_ messenger: VIMessenger, didEditMessage event: VIMessageEvent) {
        delegate?.didReceive(messageEvent: event)
    }
    
    func messenger(_ messenger: VIMessenger, didSendMessage event: VIMessageEvent) {
        delegate?.didReceive(messageEvent: event)
    }
    
    func messenger(_ messenger: VIMessenger, didRemoveMessage event: VIMessageEvent) {
        delegate?.didReceive(messageEvent: event)
    }
    
    func messenger(_ messenger: VIMessenger, didCreateConversation event: VIConversationEvent) {
        meNeedsUpdating = true
        delegate?.didReceive(conversationEvent: event)
    }
    
    func messenger(_ messenger: VIMessenger, didRemoveConversation event: VIConversationEvent) {
        meNeedsUpdating = true
        delegate?.didReceive(conversationEvent: event)
    }
    
    func messenger(_ messenger: VIMessenger, didEditConversation event: VIConversationEvent) {
        delegate?.didReceive(conversationEvent: event)
    }
    
    func messenger(_ messenger: VIMessenger, didReceiveTypingNotification event: VIConversationServiceEvent) {
        if event.imUserId == me?.imId { return }
        delegate?.didReceive(serviceEvent: event)
    }
    
    func messenger(_ messenger: VIMessenger, didReceiveReadConfirmation event: VIConversationServiceEvent) {
        if event.imUserId == me?.imId { return }
        delegate?.didReceive(serviceEvent: event)
    }
    
    func messenger(_ messenger: VIMessenger, didEditUser event: VIUserEvent) {
        delegate?.didReceive(userEvent: event)
    }
}

// MARK: - Extensions -
fileprivate extension NSError {
    static func buildError(from errorEvent: VIErrorEvent) -> NSError {
        let description: String = errorEvent.errorDescription
        return NSError(domain: errorDomain,
                       code: errorEvent.errorCode,
                       userInfo: [NSLocalizedDescriptionKey: description])
    }
}

File: client-android-kotlin/messaging/src/main/java/com/voximplant/demos/messaging/repository/remote/VoximplantService.kt
package com.voximplant.demos.messaging.repository.remote

import com.voximplant.demos.messaging.repository.utils.CustomData
import com.voximplant.demos.messaging.utils.payload.Payload
import com.voximplant.sdk.messaging.*
import kotlin.Result.Companion.failure
import kotlin.Result.Companion.success

class VoximplantService(private val messenger: IMessenger) : IMessengerListener {
    private var listener: VoximplantServiceListener? = null

    val myUsername: String?
        get() = if (messenger.me.split(".").size == 2) {
            messenger.me
        } else {
            messenger.me.substringBeforeLast(".")
        }

    init {
        this.messenger.addMessengerListener(this)
    }

    fun setListener(listener: VoximplantServiceListener) {
        this.listener = listener
    }

    //region Users
    fun requestUser(imID: Long, completion: (Result<IUser>) -> Unit) {
        messenger.getUser(imID, object : IMessengerCompletionHandler<IUserEvent> {
            override fun onSuccess(event: IUserEvent) {
                completion(success(event.user))
            }

            override fun onError(event: IErrorEvent) {
                completion(failure(buildError(event)))
            }
        })
    }

    fun requestUser(username: String, completion: (Result<IUser>) -> Unit) {
        messenger.getUser(username, object : IMessengerCompletionHandler<IUserEvent> {
            override fun onSuccess(event: IUserEvent) {
                completion(success(event.user))
            }

            override fun onError(event: IErrorEvent) {
                completion(failure(buildError(event)))
            }
        })
    }

    fun requestUsers(imIDs: List<Long>, completion: (Result<List<IUser>>) -> Unit) {
        messenger.getUsersByIMId(imIDs, object : IMessengerCompletionHandler<List<IUserEvent>> {
            override fun onSuccess(userEvents: List<IUserEvent>) {
                completion(success(userEvents.map { it.user }))
            }

            override fun onError(errorEvent: IErrorEvent) {
                completion(failure(buildError(errorEvent)))
            }
        })
    }

    fun editUser(customData: CustomData, completion: (Result<IUserEvent>) -> Unit) {
        messenger.editUser(customData, null, object : IMessengerCompletionHandler<IUserEvent> {
            override fun onSuccess(userEvent: IUserEvent) {
                completion(success(userEvent))
            }

            override fun onError(errorEvent: IErrorEvent) {
                completion(failure(buildError(errorEvent)))
            }
        })
    }
    //endregion

    //region Conversations
    fun createConversation(
        config: ConversationConfig,
        completion: (Result<IConversationEvent>) -> Unit
    ) {
        messenger.createConversation(
            config,
            object : IMessengerCompletionHandler<IConversationEvent> {
                override fun onSuccess(conversationEvent: IConversationEvent) {
                    completion(success(conversationEvent))
                }

                override fun onError(errorEvent: IErrorEvent) {
                    completion(failure(buildError(errorEvent)))
                }
            })
    }

    fun requestMultipleConversations(
        uuids: List<String>,
        completion: (Result<List<IConversation>>) -> Unit
    ) {
        messenger.getConversations(
            uuids,
            object : IMessengerCompletionHandler<List<IConversationEvent>> {
                override fun onSuccess(conversationEvents: List<IConversationEvent>) {
                    completion(success(conversationEvents.map { it.conversation }))
                }

                override fun onError(errorEvent: IErrorEvent) {
                    completion(failure(buildError(errorEvent)))
                }
            })
    }

    fun recreateMessage(uuid: String, conversationUUID: String): IMessage? =
        messenger.recreateMessage(uuid, conversationUUID, null, null, 0)


    fun recreateConversation(
        uuid: String,
        config: ConversationConfig,
        lastSequence: Long?
    ): IConversation? =
        messenger.recreateConversation(config, uuid, lastSequence ?: 0, 0, 0)


    fun addParticipants(
        participants: List<ConversationParticipant>,
        conversation: IConversation,
        completion: (Result<IConversationEvent>) -> Unit
    ) {
        conversation.addParticipants(
            participants,
            object : IMessengerCompletionHandler<IConversationEvent> {
                override fun onSuccess(conversationEvent: IConversationEvent) {
                    completion(success(conversationEvent))
                }

                override fun onError(errorEvent: IErrorEvent) {
                    completion(failure(buildError(errorEvent)))
                }
            })
    }

    fun editParticipants(
        participants: List<ConversationParticipant>,
        conversation: IConversation,
        completion: (Result<IConversationEvent>) -> Unit
    ) {
        conversation.editParticipants(
            participants,
            object : IMessengerCompletionHandler<IConversationEvent> {
                override fun onSuccess(conversationEvent: IConversationEvent) {
                    completion(success(conversationEvent))
                }

                override fun onError(errorEvent: IErrorEvent) {
                    completion(failure(buildError(errorEvent)))
                }
            })
    }

    fun removeParticipants(
        participants: List<ConversationParticipant>,
        conversation: IConversation,
        completion: (Result<IConversationEvent>) -> Unit
    ) {
        conversation.removeParticipants(
            participants,
            object : IMessengerCompletionHandler<IConversationEvent> {
                override fun onSuccess(conversationEvent: IConversationEvent) {
                    completion(success(conversationEvent))
                }

                override fun onError(errorEvent: IErrorEvent) {
                    completion(failure(buildError(errorEvent)))
                }
            })
    }

    fun updateConversation(
        conversation: IConversation,
        completion: (Result<IConversationEvent>) -> Unit
    ) {
        conversation.update(object : IMessengerCompletionHandler<IConversationEvent> {
            override fun onSuccess(conversationEvent: IConversationEvent) {
                completion(success(conversationEvent))
            }

            override fun onError(errorEvent: IErrorEvent) {
                completion(failure(buildError(errorEvent)))
            }
        })
    }

    fun leaveConversation(uuid: String, completion: (Result<IConversationEvent>) -> Unit) {
        messenger.leaveConversation(uuid, object : IMessengerCompletionHandler<IConversationEvent> {
            override fun onSuccess(conversationEvent: IConversationEvent) {
                completion(success(conversationEvent))
            }

            override fun onError(errorEvent: IErrorEvent) {
                completion(failure(buildError(errorEvent)))
            }
        })
    }

    //endregion

    fun requestMessengerEvents(
        conversation: IConversation,
        numberOfEvents: Int,
        sequence: Long,
        completion: (Result<List<IMessengerEvent>>) -> Unit
    ) {
        conversation.retransmitEventsTo(
            sequence,
            numberOfEvents,
            object : IMessengerCompletionHandler<IRetransmitEvent> {
                override fun onSuccess(retransmitEvents: IRetransmitEvent) {
                    completion(success(retransmitEvents.events))
                }

                override fun onError(errorEvent: IErrorEvent) {
                    completion(failure(buildError(errorEvent)))
                }
            })
    }

    fun sendMessage(
        text: String,
        payload: Payload?, conversation: IConversation,
        completion: (Result<IMessageEvent>) -> Unit
    ) {
        conversation.sendMessage(
            text,
            payload,
            object : IMessengerCompletionHandler<IMessageEvent> {
                override fun onSuccess(messageEvent: IMessageEvent) {
                    completion(success(messageEvent))
                }

                override fun onError(errorEvent: IErrorEvent) {
                    completion(failure(buildError(errorEvent)))
                }
            })
    }

    fun markAsRead(sequence: Long, conversation: IConversation) {
        conversation.markAsRead(sequence, null)
    }

    fun sendTyping(conversation: IConversation) {
        conversation.typing(null)
    }

    fun removeMessage(message: IMessage, completion: (Result<IMessageEvent>) -> Unit) {
        message.remove(object : IMessengerCompletionHandler<IMessageEvent> {
            override fun onSuccess(messageEvent: IMessageEvent) {
                completion(success(messageEvent))
            }


            override fun onError(errorEvent: IErrorEvent) {
                completion(failure(buildError(errorEvent)))
            }
        })
    }

    fun editMessage(message: IMessage, text: String, completion: (Result<IMessageEvent>) -> Unit) {
        message.update(text, null, object : IMessengerCompletionHandler<IMessageEvent> {
            override fun onSuccess(messageEvent: IMessageEvent) {
                completion(success(messageEvent))
            }

            override fun onError(errorEvent: IErrorEvent) {
                completion(failure(buildError(errorEvent)))
            }
        })
    }

    //region IMessengerListener
    override fun onSendMessage(event: IMessageEvent) {
        listener?.onMessageEvent(event)
    }

    override fun onEditMessage(event: IMessageEvent) {
        listener?.onMessageEvent(event)
    }

    override fun onRemoveMessage(event: IMessageEvent) {
        listener?.onMessageEvent(event)
    }

    override fun onCreateConversation(event: IConversationEvent) {
        listener?.onConversationEvent(event)
    }

    override fun onRemoveConversation(event: IConversationEvent) {
        listener?.onConversationEvent(event)
    }

    override fun onEditConversation(event: IConversationEvent) {
        listener?.onConversationEvent(event)
    }

    override fun onTyping(event: IConversationServiceEvent) {
        listener?.onServiceEvent(event)
    }

    override fun isRead(event: IConversationServiceEvent) {
        listener?.onServiceEvent(event)
    }

    override fun onEditUser(event: IUserEvent) {
        listener?.onUserEvent(event)
    }

    override fun onGetUser(p0: IUserEvent?) {

    }

    override fun onGetPublicConversations(p0: IConversationListEvent?) {}

    override fun onSetStatus(p0: IStatusEvent?) {}

    override fun onGetConversation(p0: IConversationEvent?) {}

    override fun onGetSubscriptionList(p0: ISubscriptionEvent?) {}

    override fun onUnsubscribe(p0: ISubscriptionEvent?) {}

    override fun onRetransmitEvents(p0: IRetransmitEvent?) {}

    override fun onSubscribe(p0: ISubscriptionEvent?) {}

    override fun onError(p0: IErrorEvent?) {}
//endregion

    private fun buildError(event: IErrorEvent): Error {
        return Error(event.errorDescription)
    }
}